### PR TITLE
fix cross grid iteration domain in corss grid reduction

### PR DESCRIPTION
**Pull request by liqiangxl**
- In the inner reduction scheduler, when we encounter a cross-grid situation, we employ `gdimy` to parallelize the iteration domain, as `gdimx` is allocated for the reduction domain. 

- The gdimy value is set equal to `godim`, computed as `godim = ceilDiv(total_iteration_numel, bdimy * iter_unroll_factor);`. This value is subsequently assessed to determine if we must invoke `split_grid_dim_iter_dom_outer` to prevent it from surpassing the limit set by `scheduler_utils::y_grid_limit`. 

- Currently, in TOT, `iter_unroll_factor` is changed without update `godim`. This oversight can result in requesting a `gdimy` count exceeding the `scheduler_utils::y_grid_limit`. The newly added test case fails without the fix in this PR with the err msg (to reproduce, needs to remove assertions in the test):
```
 C++ exception with description "gdimy() > 0 && gdimy() <= 65535 INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/executor_params.cpp":27, please report a bug to PyTorch. Invalid number of blocks in y direction: 65543
Exception raised from assertValid at /opt/pytorch/nvfuser/csrc/executor_params.cpp:27 (most recent call first):
 ```

### CODE REVIEW

**Comment by naoyam** [comment]
I could check by myself, but it would be really helpful if you could add more info as part of the PR description. Is this a correctness fix or a performance fix? I see you add a new test. Does that test fail with TOT?

File: test/test_gpu3.cpp [hunk 1]
```diff
@@ -9763,6 +9763,53 @@ TEST_F(NVFuserTest, IndexDataTypePromotion) {
...
+  auto t1 = sum(t0, {1});
+  fusion.addInput(t0);
+  fusion.addOutput(t1);
+
+  auto options =
+      at::TensorOptions().dtype(data_type_to_aten(dtype)).device(at::kCUDA, 0);
+  at::Tensor aten_input = at::randn(input_shape, options);
+
+  auto reduction_params = getReductionHeuristics(&fusion, {aten_input});
+  TORCH_CHECK(reduction_params, "Reduction schedule was not generated!");
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply test/test_gpu3.cpp:9788]: changed.

File: test/test_gpu3.cpp [hunk 2]
```diff
@@ -9763,6 +9763,53 @@ TEST_F(NVFuserTest, IndexDataTypePromotion) {
...
+      reduction_params->split_grid_dim_iter_dom_outer,
+      "Expect split_grid_dim_iter_dom_outer to be true!");
+  scheduleReduction(&fusion, *reduction_params);
+
+  auto lparams = reduction_params->lparams;
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {aten_input}, lparams);
+  auto cg_outputs = fe.runFusion({aten_input}, lparams);
+
+  auto aten_outputs = aten_input.sum({1});
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply test/test_gpu3.cpp:9802]: accepted.

File: test/test_gpu3.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (naoyam) [reply test/test_gpu3.cpp:9788]: There's still one

File: test/test_gpu3.cpp [hunk 3]
```diff
@@ -9763,6 +9763,49 @@ TEST_F(NVFuserTest, IndexDataTypePromotion) {
...
+  fusion.addInput(t0);
+  fusion.addOutput(t1);
+
+  auto options =
+      at::TensorOptions().dtype(data_type_to_aten(dtype)).device(at::kCUDA, 0);
+  at::Tensor aten_input = at::randn(input_shape, options);
+
+  auto reduction_params = getReductionHeuristics(&fusion, {aten_input});
+  TORCH_CHECK(reduction_params, "Reduction schedule was not generated!");
+  ASSERT_TRUE(reduction_params->split_grid_dim_inner_reduction);
```
Comment (naoyam) [inline test/test_gpu3.cpp:9789]: Please add an error message

Comment (liqiangxl) [reply test/test_gpu3.cpp]: added

File: test/test_gpu3.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (liqiangxl) [reply test/test_gpu3.cpp]: changed.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**